ddtrace/internal/module.py

```diff
@@ -0,0 +1,410 @@
+        if path is None:
+            raise ValueError("Module origin %s cannot be resolved", origin)
+
+        if path not in sys.modules._hook_map:
+            raise ValueError("No hooks registered for origin %s" % origin)
+
+        for i, (_, a) in enumerate(sys.modules._hook_map[path]):
+            if isinstance(a, list) and arg in a:
+                # DEV: This comes from the knowledge that sometimes the argument
+                # of the hook can be a list.
+                a.remove(arg)
+                if not a:
+                    sys.modules._hook_map[path].pop(i)
+                return
+            elif a is arg:
+                sys.modules._hook_map[path].pop(i)
+                return
+        raise ValueError("No hook registered for origin %s with argument %r" % (origin, arg))
+
+    @classmethod
+    def register_module_hook(cls, module, hook, arg):
+        # type: (str, HookType, Any) -> None
+        """Register a hook to be called when the module with the given name is
+        imported.
+
+        The hook will be called with the module object and the given argument.
+        """
+        assert isinstance(sys.modules, cls), "%r is installed" % cls
+
+        log.debug("Registering hook '%r' on module '%s'", hook, module)
+        sys.modules._hook_map[module].append((hook, arg))
+        try:
+            module_object = sys.modules[module]
+        except KeyError:
+            # The module is not loaded yet. Nothing more we can do.
+            return
+
+        # The module was already imported so we invoke the hook straight-away
+        log.debug("Calling hook '%r' on already imported module '%s'", hook, module)
+        hook(module_object, arg)
+
+    @classmethod
+    def unregister_module_hook(cls, module, arg):
+        # type: (str, Any) -> None
+        """Unregister the hook registered with the given module name and
+        argument.
+        """
+        assert isinstance(sys.modules, cls), "%r is installed" % cls
+
+        if module not in sys.modules._hook_map:
+            raise ValueError("No hooks registered for module %s" % module)
+
+        for i, (_, a) in enumerate(sys.modules._hook_map[module]):
+            if isinstance(a, list) and arg in a:
+                # DEV: This comes from the knowledge that sometimes the argument
+                # of the hook can be a list.
+                a.remove(arg)
+                if not a:
+                    sys.modules._hook_map[module].pop(i)
+                return
+            elif a is arg:
+                sys.modules._hook_map[module].pop(i)
+                return
+        raise ValueError("No hook registered for module %s with argument %r" % (module, arg))
+
+    @classmethod
+    def install(cls, on_run_module=None):
+        # type: (Optional[Callable[[Callable[..., Any]], Callable[..., Any]]]) -> None
+        """Install the module watchdog.
+
+        The optional `on_run_module` is a callable that implements a wrapper
```

Comment:
The runpy support amounts to just patching `runpy._run_code`. It might be wasteful if somebody wanted to use a ModuleWatchdog just for this, so perhaps we can take this out into its own function/class?

Reply:
Yeah maybe just make it its own function in this module? The logic seems pretty distinct to the rest of the class.

It's also easy enough to change later so I'm ok keeping it.

Reply:
I've refactored this similarly to `forksafe`